### ISSUE TITLE
xgb.model.dt.tree.R fails when tree has only 1 node that is not a Leaf

### ISSUE THREAD

**Issue by picarus** [dmlc/xgboost#2191]
For bugs or installation issues, please provide the following information.
The more information you provide, the more easily we will be able to offer
help and advice.

## Environment info
Operating System: Ubuntu

Compiler: gcc (Ubuntu 5.4.0-6ubuntu1~16.04.4) 5.4.0 20160609

Package used (python/R/jvm/C++): R

`xgboost` version used: 0.6-4

If you are using R package, please provide

1. The R `sessionInfo()`

R version 3.3.2 (2016-10-31)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 16.04.1 LTS

locale:
 [1] LC_CTYPE=en_AU.UTF-8       LC_NUMERIC=C
 [3] LC_TIME=en_AU.UTF-8        LC_COLLATE=en_AU.UTF-8
 [5] LC_MONETARY=en_AU.UTF-8    LC_MESSAGES=en_AU.UTF-8
 [7] LC_PAPER=en_AU.UTF-8       LC_NAME=C
 [9] LC_ADDRESS=C               LC_TELEPHONE=C
[11] LC_MEASUREMENT=en_AU.UTF-8 LC_IDENTIFICATION=C

attached base packages:
[1] parallel  stats     graphics  grDevices utils     datasets  methods
[8] base

other attached packages:
 [1] stringi_1.1.3     ROCR_1.0-7        gplots_3.0.1      doMC_1.3.4
 [5] iterators_1.0.8   foreach_1.4.3     ggplot2_2.2.1     magrittr_1.5
 [9] data.table_1.10.4 xgboost_0.6-4

loaded via a namespace (and not attached):
 [1] Rcpp_0.12.9        munsell_0.4.3      colorspace_1.3-2   lattice_0.20-34
 [5] plyr_1.8.4         caTools_1.17.1     tools_3.3.2        grid_3.3.2
 [9] gtable_0.2.0       KernSmooth_2.23-15 gtools_3.5.0       lazyeval_0.2.0
[13] assertthat_0.1     tibble_1.2         Matrix_1.2-7.1     bitops_1.0-6
[17] codetools_0.2-15   gdata_2.17.0       scales_0.4.1

2. The command to install `xgboost` if you are not installing from source

Xgboost is an installed by an admin. I may get the info if really needed but could take a while.

## Steps to reproduce

1. Obtain a model from xgb.train that has only one node that is not a leaf (one is provided as attachment). Feature Names are irrelevant and cannot be provided for confidentiality issues. The model has 388 features if that helps.
2. Try to obtain the top features using xgb.importance
3. The method calls xgb.model.dt.tree.R and fails with error:

## What have you tried?

1. xtr[3:5] works while xtr[,3:5] produces the error. xtr is a character vector when a data.frame is expected
2.  the error message is Error in xtr[, 3:5] : incorrect number of dimensions
3. I have debugged the code, by running the source with param replacement and copy and paste. 
[xgb-issue-model.Rd.zip](https://github.com/dmlc/xgboost/files/918056/xgb-issue-model.Rd.zip)

**Comment by khotilov** [comment]
Thanks for the report. My hunch is that the line 
```R
xtr <- stri_match_first_regex(t, rx)[, c(2,3,5,6,7,8,10)]
```
needs to be changed to 
```R
xtr <- stri_match_first_regex(t, rx)[, c(2,3,5,6,7,8,10), drop=FALSE]
```
Could you please check if that would fix the issue? The easiest way would be to use `fix(xgb.model.dt.tree)` to temporary change the code of the function.

**Comment by picarus** [comment]
I have tried the suggested fix and still the same problem happens.
Maybe the drop happens before at stri_match_first_regex?

**Comment by khotilov** [comment]
`fix` does not change the function in a package, but only creates an edited copy in the global namespace. `xgb.importance` still calls `xgb.model.dt.tree` from the package. You need to [rebuild the package](https://xgboost.readthedocs.io/en/latest/build.html) from source after checking it out and modifying the source of xgb.model.dt.tree.R. However, you would not see much interesting in the result of `xgb.importance` from a model with a single feature split, as all the numbers would be normalized to one.

**Comment by picarus** [comment]
Yes, I understand that, and to prevent this I have also applied fix to the other function (without changing anything, adding a line feed). I would have expected that to take care of the definition issue.
I agree the model is not interesting, but I am running an stochastic search and this one has crashed my program. 
I have found other anomalies too... like xgb.importance complaining the model was not a tree model when I always use the same settings on that matter.
Thanks for your help.
